### DISCUSSION TITLE
Typed tags

### DISCUSSION THREAD

**Discussion by ghost** [junit-team/junit-framework#3581] [General]
There is a similar discussion in this [feature request](https://github.com/junit-team/junit5/issues/1160).

Meta-and composed annotations are a good solution for stricter tag typing but it does not solve one, as it seems to me, common case.

Let's say we have a test classification consisting of 10+ sometimes changing values:
1. red
2. green
3. ...

That means we need either 10+ meta annotations:
1. `@Red`
2. `@Green`
3. ...

or, for example, use @Tag annotation and a storage class for 10+ constants:

```
public class Color {
    public static final String RED = "red";
    public static final String GREEN = "greed";
    ...
```
1. `Tag(Color.RED)`
2. `Tag(Color.GREEN)`
3. ...

A better approach would be the ability to pass, for example, an enum to @Tag annotation with the ability to get the desired text from it.

What do you think?

**Comment by sbrannen** [comment]
> A better approach would be the ability to pass, for example, an enum to @tag annotation with the ability to get the desired text from it.

That's unfortunately not possible in Java.

Although an annotation can have an attribute with an enum type, it must be a concrete `Enum` subtype.

In other words, there is no way to declare an annotation attribute that can be assigned _any_ enum constant value.
